modules/ssh/ssh.go

```diff
@@ -10,178 +9,179 @@ import (
+	if err == nil {
+		return 0
 	}
-	return cmd[i:]
-}
 
-func handleServerConn(keyID string, chans <-chan ssh.NewChannel) {
-	for newChan := range chans {
-		if newChan.ChannelType() != "session" {
-			err := newChan.Reject(ssh.UnknownChannelType, "unknown channel type")
-			if err != nil {
-				log.Error("Error rejecting channel: %v", err)
-			}
-			continue
-		}
+	exitErr, ok := err.(*exec.ExitError)
+	if !ok {
+		return 1
+	}
 
-		ch, reqs, err := newChan.Accept()
-		if err != nil {
-			log.Error("Error accepting channel: %v", err)
-			continue
+	waitStatus, ok := exitErr.Sys().(syscall.WaitStatus)
+	if !ok {
+		// This is a fallback and should at least let us return something useful
+		// when running on Windows, even if it isn't completely accurate.
+		if exitErr.Success() {
+			return 0
 		}
 
-		go func(in <-chan *ssh.Request) {
-			defer func() {
-				if err = ch.Close(); err != nil {
-					log.Error("Close: %v", err)
-				}
-			}()
-			for req := range in {
-				payload := cleanCommand(string(req.Payload))
-				switch req.Type {
-				case "exec":
-					cmdName := strings.TrimLeft(payload, "'()")
-					log.Trace("SSH: Payload: %v", cmdName)
-
-					args := []string{"serv", "key-" + keyID, "--config=" + setting.CustomConf}
-					log.Trace("SSH: Arguments: %v", args)
-					cmd := exec.Command(setting.AppPath, args...)
-					cmd.Env = append(
-						os.Environ(),
-						"SSH_ORIGINAL_COMMAND="+cmdName,
-						"SKIP_MINWINSVC=1",
-					)
-
-					stdout, err := cmd.StdoutPipe()
-					if err != nil {
-						log.Error("SSH: StdoutPipe: %v", err)
-						return
-					}
-					stderr, err := cmd.StderrPipe()
-					if err != nil {
-						log.Error("SSH: StderrPipe: %v", err)
-						return
-					}
-					input, err := cmd.StdinPipe()
-					if err != nil {
-						log.Error("SSH: StdinPipe: %v", err)
-						return
-					}
-
-					// FIXME: check timeout
-					if err = cmd.Start(); err != nil {
-						log.Error("SSH: Start: %v", err)
-						return
-					}
-
-					err = req.Reply(true, nil)
-					if err != nil {
-						log.Error("SSH: Reply: %v", err)
-					}
-					go func() {
-						_, err = io.Copy(input, ch)
-						if err != nil {
-							log.Error("SSH: Copy: %v", err)
-						}
-					}()
-					_, err = io.Copy(ch, stdout)
-					if err != nil {
-						log.Error("SSH: Copy: %v", err)
-					}
-					_, err = io.Copy(ch.Stderr(), stderr)
-					if err != nil {
-						log.Error("SSH: Copy: %v", err)
-					}
-
-					if err = cmd.Wait(); err != nil {
-						log.Error("SSH: Wait: %v", err)
-						return
-					}
-
-					_, err = ch.SendRequest("exit-status", false, []byte{0, 0, 0, 0})
-					if err != nil {
-						log.Error("SSH: SendRequest: %v", err)
-					}
-					return
-				default:
-				}
-			}
-		}(reqs)
+		return 1
 	}
+
+	return waitStatus.ExitStatus()
 }
 
-func listen(config *ssh.ServerConfig, host string, port int) {
-	listener, err := net.Listen("tcp", host+":"+com.ToStr(port))
+func shellEscape(args []string) string {
```

Comment:
This function is unnecessary

Reply:
removed. Thanks :)

Reply:
I don't believe this is true. This function was added so spaces (and other characters) would be properly passed through. The problem is that we're taking a slice of args and trying to fit it into a single value (an environment variable). shellEscape was an attempt to encode problematic values so they could be properly pulled out of shellEscape as separate args.